vcd/resource_vcd_edgegateway_test.go

```diff
@@ -0,0 +1,155 @@
+				),
+			},
+		},
+	})
+}
+
+func TestAccVcdEdgeGatewayComplex(t *testing.T) {
+	var (
+		edgeGatewayName       string = "TestEdgeGatewayComplex"
+		edgeGatewayVcdName    string = "test_edge_gateway_complex"
+		newExternalNetwork    string = "TestExternalNetwork"
+		newExternalNetworkVcd string = "test_external_network"
+	)
+
+	// String map to fill the template
+	var params = StringMap{
+		"Org":                   testConfig.VCD.Org,
+		"Vdc":                   testConfig.VCD.Vdc,
+		"EdgeGateway":           edgeGatewayName,
+		"EdgeGatewayVcd":        edgeGatewayVcdName,
+		"ExternalNetwork":       testConfig.Networking.ExternalNetwork,
+		"Tags":                  "gateway",
+		"NewExternalNetwork":    newExternalNetwork,
+		"NewExternalNetworkVcd": newExternalNetworkVcd,
+		"Type":                  testConfig.Networking.ExternalNetworkPortGroupType,
+		"PortGroup":             testConfig.Networking.ExternalNetworkPortGroup,
+		"Vcenter":               testConfig.Networking.Vcenter,
+	}
+	configText := templateFill(testAccEdgeGatewayComplex, params)
+	if vcdShortTest {
+		t.Skip(acceptanceTestsSkipped)
+		return
+	}
+	debugPrintf("#[DEBUG] CONFIGURATION: %s", configText)
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckVcdEdgeGatewayDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: configText,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"vcd_edgegateway."+edgeGatewayName, "default_gateway", newExternalNetworkVcd),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckVcdEdgeGatewayDestroy(s *terraform.State) error {
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "vcd_edgegateway" {
+			continue
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+const testAccEdgeGatewayBasic = `
+resource "vcd_edgegateway" "{{.EdgeGateway}}" {
+  org                 = "{{.Org}}"
+  vdc                 = "{{.Vdc}}"
+  name                = "{{.EdgeGatewayVcd}}"
+  description         = "Description"
+  backing_config      = "compact"
+  default_gateway     = "{{.ExternalNetwork}}"
+
+  external_networks   = [ "{{.ExternalNetwork}}" ]
+}
+`
+
+const testAccEdgeGatewayComplex = `
+
+resource "vcd_external_network" "{{.NewExternalNetwork}}" {
+  name        = "{{.NewExternalNetworkVcd}}"
+  description = "Test External Network"
+
+  vsphere_network {
+    vcenter = "{{.Vcenter}}"
+    name    = "{{.PortGroup}}"
+    type    = "{{.Type}}"
+  }
+
+  ip_scope {
+    gateway      = "192.168.30.49"
+    netmask      = "255.255.255.240"
+    dns1         = "192.168.0.164"
+    dns2         = "192.168.0.196"
+    dns_suffix   = "company.biz"
+
+    static_ip_pool {
+      start_address = "192.168.30.51"
+      end_address   = "192.168.30.62"
+    }
+  }
+
+  retain_net_info_across_deployments = "false"
+}
+
+resource "vcd_edgegateway" "{{.EdgeGateway}}" {
+  org                 = "{{.Org}}"
```

Comment:
Test doesn't cover all parameters

Reply:
We can't cover all parameters as there are limitations on some of them. 
1. advanced networking can only be "true" on 9.7+
2. HA can be set, but we don't have the needed info to follow up on that setting
3. ditto for distributed routing.

That said, I am reluctant to add more cases, as the creation of edge gateway takes up several minutes. I could add a test without default gateway and using the "full" configuration but this would double the execution time for this test already. Writing the code to do it would be quick, but the execution time for the whole suite would go up considerably.

Reply:
I am confused now...
1. I though `advanced` is same as checkbox in 9.1-9.7 versions "Create as advanced gateway". Also nothing mentioned in documentation or warning in running resource if it is only 9.7+ version
2. HA as far as I got is "Enable High availability" in UI
3. And this one is possible to as checkbox "Enable Distributed networking" when ""Create as advanced gateway" is checked. - So here we miss validation I think.

Reply:
And I agree that using "Advanced" make test more longer. Question how we sill cover that..

Reply:
>     1. I though `advanced` is same as checkbox in 9.1-9.7 versions "Create as advanced gateway". Also nothing mentioned in documentation or warning in running resource if it is only 9.7+ version

There is no checkbox for advanced in 9.7, and as mentioned in the issue, this property is ON by default, regardless of the input we pass.

>     2. HA as far as I got is "Enable High availability" in UI

Yes. And we can check it. The only trouble is, if we want to make a meaningful test, we should have one with and one without. And this comes back to the question of test length.

>     3. And this one is possible to as checkbox "Enable Distributed networking" when ""Create as advanced gateway" is checked. - So here we miss validation I think.

The validation is done in go-vcloud-director.

Reply:
Thank you for answers.

Reply:
So minimum what we should cover case for 9.7 in documentation. Or even warning in code.

Reply:
The documentation mentions it, and the resource will return an error if you set `advanced` as false for 9.7.

Reply:
If we need to choose, it feels more actual to test with `advanced` these days.

Reply:
What we are saying here is that we are forcing users to set `advanced=true` to true for 9.7.
The change that I mentioned is about making the `advanced` property mandatory, because otherwise we can't set a sensible default.

Reply:
We can make it optional with default set to `true`, this way users with vCD 9.7+ can omit the field, while the users of older versions will have to set advanced to false. In other words, the field will automatically become not used with newer vCD versions, which is good as the less fields to set the better.

Reply:
Changed to optional with default = `true`